resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/Client.scala

```diff
@@ -16,256 +16,130 @@
+    * Driver Pod, and Kubernetes Resources
+    *
+    */
   def run(): Unit = {
-    val arguments = (pythonResource map {p => p.arguments}).getOrElse(appArgs)
-    val driverCustomLabels = ConfigurationUtils.combinePrefixedKeyValuePairsWithDeprecatedConf(
-      sparkConf,
-      KUBERNETES_DRIVER_LABEL_PREFIX,
-      KUBERNETES_DRIVER_LABELS,
-      "label")
-    require(!driverCustomLabels.contains(SPARK_APP_ID_LABEL), s"Label with key " +
-        s" $SPARK_APP_ID_LABEL is not allowed as it is reserved for Spark bookkeeping" +
-        s" operations.")
-
-    val driverCustomAnnotations = ConfigurationUtils.combinePrefixedKeyValuePairsWithDeprecatedConf(
-      sparkConf,
-      KUBERNETES_DRIVER_ANNOTATION_PREFIX,
-      KUBERNETES_DRIVER_ANNOTATIONS,
-      "annotation")
-    require(!driverCustomAnnotations.contains(SPARK_APP_NAME_ANNOTATION),
-        s"Annotation with key $SPARK_APP_NAME_ANNOTATION is not allowed as it is reserved for" +
-        s" Spark bookkeeping operations.")
-    val allDriverLabels = driverCustomLabels ++ Map(
-        SPARK_APP_ID_LABEL -> kubernetesAppId,
-        SPARK_ROLE_LABEL -> SPARK_POD_DRIVER_ROLE)
-
-    val driverExtraClasspathEnv = driverExtraClasspath.map { classPath =>
-      new EnvVarBuilder()
-        .withName(ENV_SUBMIT_EXTRA_CLASSPATH)
-        .withValue(classPath)
-        .build()
+    // Set new metadata and a new spec so that submission steps can use PodBuilder#editMetadata()
+    // and/or PodBuilder#editSpec() safely.
+    val basePod = new PodBuilder().withNewMetadata().endMetadata().withNewSpec().endSpec().build()
+    var currentDriverSpec = KubernetesDriverSpec(
+      driverPod = basePod,
+      driverContainer = new ContainerBuilder().build(),
+      driverSparkConf = submissionSparkConf.clone(),
+      otherKubernetesResources = Seq.empty[HasMetadata])
+    // submissionSteps contain steps necessary to take, to resolve varying
+    // client arguments that are passed in, created by orchestrator
+    for (nextStep <- submissionSteps) {
+      currentDriverSpec = nextStep.prepareSubmission(currentDriverSpec)
     }
-    val driverCpuQuantity = new QuantityBuilder(false)
-      .withAmount(driverCpuCores)
-      .build()
-    val driverMemoryQuantity = new QuantityBuilder(false)
-      .withAmount(s"${driverMemoryMb}M")
-      .build()
-    val driverMemoryLimitQuantity = new QuantityBuilder(false)
-      .withAmount(s"${driverContainerMemoryWithOverhead}M")
-      .build()
-    val driverContainer = new ContainerBuilder()
-      .withName(DRIVER_CONTAINER_NAME)
-      .withImage(driverDockerImage)
-      .withImagePullPolicy(dockerImagePullPolicy)
-      .addToEnv(driverExtraClasspathEnv.toSeq: _*)
+    val resolvedDriverJavaOpts = currentDriverSpec
+      .driverSparkConf
+      // We don't need this anymore since we just set the JVM options on the environment
+      .remove(org.apache.spark.internal.config.DRIVER_JAVA_OPTIONS)
```

Comment:
The problem with filtering the JVM options in a step is that it's not obvious just from the context of the step itself why the step is doing that. Putting the filtering right next to the application of the JVM options to the container flows better.